# SISTEMAS DE RECOMENDACIÓN

Los sistemas de recomendación son modelos cuyo objetivo es proporcionar, como su nombre indica, una recomendación de un producto basada en la experiencia el consumidor o considerando en conjunto la experiencia del consumidor junto con otros.

Tradicionalmente, los sistemas de recomendación se han basado en la creación de rankings, de modo que se recomienda aquello que más ha votado una comunidad de usuarios.

Gracias a la evolución de los algoritmos de inteligencia artificial, se plantea la posiblidad de aplicar métodos más complejos para la obtención de recomendaciones más precisas y personalizadas.

Así surgen, entre otros, los llamados sistemas colaborativos, que intentan obtener un perfilado o tratar de asignar recomendaciones basadas en perfiles similares.

Entre estos últimos podemos destacar los algoritmos basados en distancias y las descomposiciones matriciales. 

En el ejemplo que se va a desarrollar se va a mostrar un ejemplo tanto de de sistema basado en un ranking genérico y un ejemplo más detallado con un sistema de distancias.

# ANÁLISIS DEL DATASET

Vamos a utilizar un dataset de la plataforma Kaggle que consiste en un conjunto de películas y usuarios los cuales han puntuado entre 0 y 5 las películas que han visto.

Los datos se componen de dos archivos:
- ratings_small.csv con la relación usuario - identificador de película - puntuación
- movies_medatada.csv donde hay información asociada de la película como su nombre y género.

In [2]:
### Carga de librerías genéricas

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings

plt.style.use("ggplot")
warnings.filterwarnings("ignore")

In [3]:
##Direcciones de los archivos

camino="D:\\J\\Big data\\DATAHACK\\PYTHON\\006-_Caso_segmentacion\\Recomendador peliculas\\"
archivo="ratings_small.csv"
archivo_pelis="movies_metadata.csv"

In [4]:
#Lectura del fichero con los ratings
df=pd.read_csv(camino+archivo)

In [5]:
# Información asociada a las películas
df_info_pelis=pd.read_csv(camino+archivo_pelis)

## ANÁLISIS DE LOS DATAFRAMES

Procedemos a analizar los dos daframes y su estructura, comenzando por el de los rankings

In [6]:
df.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

El usuaro se identifica con un id, así como la película- El rating es un número, veremos entre qué se mueve- El timestamp no lo vamos a utilizar, con lo que podemos eliminarlo en primer lugar

In [7]:
df.drop(columns=["timestamp"],inplace=True)

In [8]:
df["rating"].value_counts().sort_index(ascending=False)

5.0    15095
4.5     7723
4.0    28750
3.5    10538
3.0    20064
2.5     4449
2.0     7271
1.5     1687
1.0     3326
0.5     1101
Name: rating, dtype: int64

En efecto, se ve que los ratings van desde 0.5 hasta 5. Veamos cuántos usuarios distintos hay, así como películas:

In [9]:
df["userId"].unique().shape

(671,)

Hay 671 usuarios diferentes

In [10]:
### ESCRIBE TU CÓDIGO AQUÍ PARA OBTENER EL NÚMERO DE PELÍCULAS DISTINTAS

df["movieId"].unique().shape

(9066,)

Hay 9066 películas diferentes. Veamos ahora el archivo de las películas:

In [11]:
df_info_pelis.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [12]:
df_info_pelis.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count'],
      dtype='object')

De todas las columnas nos interesan la que tiene el género de la película, el id para poder cruzar, el título y la media de votos. Filtraomos el df por estos campos.

In [13]:
###TU CÓDIGO PARA QUEDARNOS SOLO CON LAS COLUMNAS:["genres","id","title","vote_average"]
### MODIFICA EL DF df_info_pelis 

df_info_pelis=df_info_pelis[["genres","id","title","vote_average"]]
df_info_pelis.head()

genres     id  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357   
4                     [{'id': 35, 'name': 'Comedy'}]  11862   

                         title  vote_average  
0                    Toy Story           7.7  
1                      Jumanji           6.9  
2             Grumpier Old Men           6.5  
3            Waiting to Exhale           6.1  
4  Father of the Bride Part II           5.7

La columna "genres" tiene una estructura un tanto complicada que habría que limpiar

In [14]:
df_info_pelis["genres"][0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

Vemos que una película se puede clasificar de diferentes maneras. Para explotarlo tenemos que ver el tipo de columna:

In [15]:
df_info_pelis["genres"].dtype

dtype('O')

Podemos detectar el género porque se ve que tienen la primera mayúscula. Así podemos filtrar la lista de modo sencillo porque python identifica este formato en textos con el método .istitle()

In [65]:
print("Hola".istitle())
print("hola".istitle())

True
False


La siguiente línea de código crea una nueva columna solo con los géneros:

In [66]:
df_info_pelis["generos"]=df_info_pelis["genres"].apply(lambda x: [y for y in x.split("'") if y.istitle()])

Sin embargo, sigue siendo poco manejable, pues el elemento de la columna es un poco "extraño":


In [73]:
##TU CÓDIGO AQUÍ VIENDO EL TIPO DE ELEMENTO DE ESTA COLUMNA.
## AQUÍ DTYPE NO VA A FUNCIONAR PORQUE TE VA A DEVOLVER ALGO GENÉRICCO TIPO OBJECT.
## SACA UN ELEMENTO DE LA COLUMNA Y UTILIZA type

type(df_info_pelis["generos"][0])

list

Tenemos listas que habría que explotar, pues parece que es compatible que una película sea de animación y de aventuras a la vez, pero ¿cuántos es el máximo número de géneros para una sola película?

In [74]:
np.max(df_info_pelis["generos"].apply(lambda x: len(x)))

8

En efecto, se ve que hay hasta 8 géneros distintos para una sola película. La siguiente línea de código nos va a indicar cuántos distintos hay:

In [39]:
set([y for x in df_info_pelis["generos"].values for y in x])

{'Action',
 'Adventure',
 'Animation',
 'Aniplex',
 'Carousel Productions',
 'Comedy',
 'Crime',
 'Documentary',
 'Drama',
 'Family',
 'Fantasy',
 'Foreign',
 'History',
 'Horror',
 'Mardock Scramble Production Committee',
 'Music',
 'Mystery',
 'Odyssey Media',
 'Pulser Productions',
 'Rogue State',
 'Romance',
 'Science Fiction',
 'Sentai Filmworks',
 'Telescene Film Group Productions',
 'The Cartel',
 'Thriller',
 'Vision View Entertainment',
 'War',
 'Western'}

La idea ahora es crear una lista con los que entedamos que son principales y después quedarnos solo con el primero. Crearemos una categría nueva con "otros" para los que no sean de ningún tipo. Sin embargo, antes vamos a limpiar el dataframe eliminando aquellas películas que no tienen ningún género asociado. La línea siguiente es la que lo hace:

In [75]:
df_info_pelis=df_info_pelis[df_info_pelis["generos"].apply(lambda x: bool(x))]

Ahora creamos la columna "genero" con el seleccionado:

In [90]:
nivel_1=['Pulser Productions',"Action","Animation","Adventure","Comedy",'Crime','Documentary','Fantasy','Music','Horror','Mystery','Romance','Science Fiction','Thriller','War','Western']

df_info_pelis["genero"]=df_info_pelis["generos"].apply(lambda x: [y if y in nivel_1 else "Otros" for y in x ][0])

¿Cuántos hay en cada tipo?

In [91]:
###Escribe a tu código. Utiliza value_counts() en la columna:

df_info_pelis["genero"].value_counts()

Otros              13045
Comedy              8869
Action              4521
Documentary         3428
Horror              2634
Crime               1707
Thriller            1680
Adventure           1522
Romance             1220
Animation           1131
Fantasy              714
Science Fiction      655
Mystery              561
Music                491
Western              458
War                  381
Name: genero, dtype: int64

In [46]:
df_info_pelis["generos"]

0         [Animation, Comedy, Family]
1        [Adventure, Fantasy, Family]
2                   [Romance, Comedy]
3            [Comedy, Drama, Romance]
4                            [Comedy]
                     ...             
45461                 [Drama, Family]
45462                         [Drama]
45463       [Action, Drama, Thriller]
45464                              []
45465                              []
Name: generos, Length: 45466, dtype: object

In [34]:
for x in df_info_pelis["generos"].values:
    for y in x:
        print(y)

Animation
Comedy
Family
Adventure
Fantasy
Family
Romance
Comedy
Comedy
Drama
Romance
Comedy
Action
Crime
Drama
Thriller
Comedy
Romance
Action
Adventure
Drama
Family
Action
Adventure
Thriller
Adventure
Action
Thriller
Comedy
Drama
Romance
Comedy
Horror
Family
Animation
Adventure
History
Drama
Action
Adventure
Drama
Crime
Drama
Romance
Crime
Comedy
Crime
Comedy
Adventure
Action
Comedy
Crime
Comedy
Thriller
Crime
Drama
Thriller
Action
Adventure
Crime
Thriller
Drama
Fantasy
Science Fiction
Thriller
Drama
Romance
Drama
Comedy
Drama
Family
Drama
Romance
Fantasy
Science Fiction
Adventure
Drama
Crime
Drama
Crime
Science Fiction
Thriller
Mystery
Romance
Adventure
Fantasy
Drama
Comedy
Family
History
Drama
Romance
Drama
Adventure
History
Drama
Family
Comedy
Family
Romance
Comedy
Drama
Romance
Drama
Drama
War
Action
Crime
Drama
History
Drama
Romance
Action
Fantasy
Fantasy
Drama
Comedy
Thriller
Drama
Romance
Crime
Mystery
Thriller
Adventure
Animation
Drama
Family
Drama
Romance
Drama
Crime
Thriller


Drama
War
Crime
Drama
Comedy
Foreign
Western
Drama
Romance
War
Comedy
Romance
War
Drama
Drama
Drama
Comedy
Drama
Romance
Mystery
Documentary
Crime
Comedy
Crime
Drama
Romance
Documentary
Music
Documentary
Drama
Comedy
Adventure
Action
Science Fiction
Adventure
Family
Fantasy
Comedy
Romance
Adventure
Action
Comedy
Science Fiction
Horror
Action
Thriller
Science Fiction
Western
Comedy
Drama
Drama
Adventure
Drama
History
War
Science Fiction
Drama
Crime
Drama
Drama
War
Western
Adventure
Action
Science Fiction
Drama
Fantasy
Romance
Thriller
Mystery
Drama
Crime
Horror
Action
Thriller
Science Fiction
Fantasy
Horror
Comedy
Adventure
Drama
Romance
Action
Drama
History
Action
Crime
Drama
Thriller
Drama
Horror
Thriller
Music
Comedy
Action
Crime
Drama
Crime
Drama
War
Adventure
Animation
Comedy
Science Fiction
Family
War
Drama
History
Action
Romance
Drama
History
Music
Comedy
Drama
Romance
Drama
Crime
Drama
Comedy
Drama
Romance
Drama
History
Drama
Science Fiction
Thriller
Action
Drama
History
War
Adv

Adventure
Thriller
Adventure
Animation
Drama
Family
Music
Comedy
Drama
Romance
History
Drama
Drama
Family
Family
Fantasy
Adventure
Science Fiction
Comedy
Drama
Family
Fantasy
Romance
Western
Action
Adventure
Thriller
War
Action
Adventure
Thriller
War
Action
Adventure
Thriller
War
Action
Adventure
Comedy
Romance
Action
Adventure
Comedy
Romance
Comedy
Drama
Science Fiction
Comedy
Drama
Science Fiction
Drama
Drama
Drama
Drama
Comedy
Thriller
Crime
Mystery
Action
Adventure
Crime
Drama
Family
Mystery
Thriller
Comedy
Comedy
Drama
Romance
Drama
Comedy
Romance
Drama
Thriller
Drama
Adventure
Drama
Action
Romance
Family
Family
Adventure
Comedy
Comedy
Romance
Drama
Action
Crime
Drama
Drama
History
War
Horror
Mystery
Science Fiction
Action
Adventure
Family
Fantasy
Adventure
Fantasy
Drama
Action
Science Fiction
Comedy
Drama
Drama
Romance
Drama
Drama
Comedy
Drama
Comedy
Drama
War
Drama
History
Thriller
Comedy
Drama
Romance
Drama
Thriller
Crime
Drama
Drama
Action
Western
Romance
Music
Drama
Drama
Com

Fantasy
Drama
Comedy
Drama
Family
Animation
Drama
Family
Science Fiction
Mystery
Action
Science Fiction
War
Drama
Music
Comedy
Action
Comedy
Comedy
Romance
Drama
Comedy
Fantasy
Music
Romance
Comedy
Music
Romance
Adventure
Family
Comedy
Western
Comedy
Music
Comedy
Comedy
Drama
Music
Drama
Action
Music
Romance
Comedy
Music
Romance
Family
Adventure
Comedy
Documentary
Drama
Music
Romance
Animation
Family
Adventure
Family
Fantasy
Romance
Drama
Comedy
Crime
Action
Comedy
Romance
Thriller
Animation
Family
Drama
Comedy
Romance
Family
Comedy
Adventure
Action
Comedy
Crime
Comedy
Drama
Comedy
Drama
Horror
Comedy
Adventure
Action
Thriller
Adventure
Action
Comedy
Western
Documentary
Music
Drama
Comedy
Horror
Mystery
Action
Drama
History
Adventure
Comedy
Drama
Fantasy
Horror
Science Fiction
Romance
Comedy
Comedy
Crime
Drama
Adventure
Action
Thriller
Drama
Thriller
Adventure
Action
Thriller
Romance
Drama
Drama
Action
Adventure
Thriller
Science Fiction
Adventure
Action
Thriller
Comedy
Drama
Drama
Roma

Thriller
Family
Comedy
Fantasy
Action
Adventure
Drama
Music
Drama
Comedy
Horror
Mystery
Thriller
Comedy
Romance
Drama
Thriller
Mystery
Drama
Comedy
Drama
Drama
Action
Comedy
Thriller
Crime
Romance
Drama
Crime
Drama
Romance
Comedy
Crime
Drama
Mystery
Thriller
Documentary
Drama
Thriller
Comedy
Romance
Drama
Action
Thriller
Comedy
Action
Comedy
Thriller
Drama
Romance
Action
Crime
Drama
Thriller
Action
Comedy
Science Fiction
Music
Documentary
Documentary
Drama
Action
Science Fiction
Thriller
Comedy
Drama
Romance
Documentary
History
Drama
Romance
Action
Drama
Western
Horror
Mystery
Drama
Family
Drama
Comedy
Romance
Comedy
Fantasy
Horror
Music
Thriller
Action
Drama
Western
Western
Horror
Comedy
Drama
Romance
Thriller
Horror
Mystery
Adventure
Comedy
Romance
Comedy
Drama
Romance
Comedy
Music
Comedy
Romance
Action
Adventure
Comedy
Crime
Action
Adventure
Drama
Drama
Romance
Action
Comedy
Romance
Crime
Drama
Romance
Thriller
Comedy
Drama
Drama
Romance
Drama
War
Comedy
Romance
Thriller
Drama
Myste

Horror
Science Fiction
Comedy
Romance
Comedy
Drama
Crime
Comedy
Mystery
Drama
Horror
Horror
Comedy
Drama
Science Fiction
Documentary
Drama
Action
Science Fiction
Action
Drama
Horror
Crime
Thriller
Comedy
Drama
Drama
History
Comedy
Crime
Drama
Mystery
Romance
Thriller
Science Fiction
Comedy
Horror
Drama
Romance
Comedy
Music
Comedy
Fantasy
Comedy
Drama
Romance
Science Fiction
Action
Adventure
Thriller
Drama
Comedy
Drama
Foreign
Drama
Drama
Action
Comedy
Drama
Romance
Science Fiction
Thriller
Comedy
Drama
Drama
Adventure
Fantasy
Action
Drama
War
Crime
Drama
Drama
Romance
Drama
History
Crime
Comedy
Romance
Animation
Adventure
Family
Action
Crime
Drama
Thriller
Thriller
Thriller
Drama
Thriller
Romance
Drama
Action
Thriller
Drama
War
Drama
Thriller
Crime
Mystery
Romance
Adventure
Drama
Action
Crime
Drama
Thriller
Comedy
Drama
Family
Romance
Comedy
Drama
Family
Fantasy
Animation
Family
Comedy
Crime
Action
Adventure
Family
Fantasy
Drama
War
Romance
Mystery
Thriller
Comedy
Action
Thriller
Horro

Drama
Thriller
Horror
Mystery
Thriller
Drama
Horror
Thriller
Comedy
Horror
Thriller
Animation
Science Fiction
Comedy
Romance
Horror
Comedy
Drama
Romance
Comedy
Drama
War
Crime
Drama
Documentary
Drama
Thriller
Crime
Drama
Action
Thriller
Drama
Drama
Science Fiction
Comedy
Romance
Foreign
Drama
Romance
Drama
Romance
Comedy
Fantasy
Science Fiction
Crime
Drama
Science Fiction
Thriller
Drama
Thriller
Crime
Comedy
Drama
Romance
Drama
Romance
Drama
Romance
Documentary
Documentary
Romance
Music
Family
Drama
Action
War
History
Drama
Adventure
Fantasy
Drama
Comedy
Drama
Family
Romance
Drama
Comedy
Romance
Comedy
Drama
Romance
Drama
Thriller
Adventure
Fantasy
Action
Drama
Romance
Comedy
War
Documentary
History
Action
Comedy
Drama
Comedy
Drama
Foreign
War
Crime
Drama
Comedy
Drama
Action
Adventure
Mystery
Science Fiction
Thriller
Adventure
Fantasy
Family
Drama
Adventure
Drama
Family
Drama
Romance
Drama
Comedy
Romance
Comedy
Documentary
Drama
Drama
Comedy
Romance
Animation
Drama
Family
Music
Action


Science Fiction
Action
Comedy
Thriller
Drama
Comedy
Drama
Mystery
Action
Thriller
Science Fiction
Adventure
Comedy
Romance
Drama
Drama
Romance
War
Comedy
Drama
Family
Romance
Crime
Drama
Horror
Mystery
Science Fiction
Thriller
Action
Comedy
Comedy
Drama
Horror
Comedy
Romance
Drama
Romance
Comedy
Romance
Comedy
Horror
Science Fiction
Drama
Thriller
Mystery
Drama
Drama
Music
Romance
Crime
Drama
Mystery
Thriller
Adventure
Family
Action
Comedy
Horror
Drama
Mystery
Horror
Science Fiction
Mystery
Thriller
Drama
Action
Comedy
Horror
Horror
Science Fiction
Thriller
Horror
Horror
Drama
Adventure
Horror
Drama
Foreign
Horror
Science Fiction
Thriller
Drama
Action
Adventure
Drama
Crime
Drama
Thriller
Comedy
Drama
Action
Drama
Mystery
Thriller
Science Fiction
Horror
Mystery
Horror
Thriller
Horror
Action
Adventure
Drama
History
Adventure
Horror
Animation
Action
Comedy
Family
Drama
Action
Thriller
Drama
Comedy
Romance
Music
Drama
Romance
Thriller
Action
Comedy
Crime
Drama
Science Fiction
Drama
Thrille

Family
Comedy
Drama
Foreign
Mystery
Thriller
Drama
Comedy
Foreign
Mystery
Horror
Thriller
Drama
Action
Drama
Thriller
Action
Comedy
Horror
Action
Comedy
Romance
Action
Thriller
Drama
Crime
Drama
Horror
Thriller
Action
Thriller
Drama
Action
Science Fiction
Comedy
Drama
Romance
Thriller
Science Fiction
Adventure
Action
Animation
Comedy
Family
Drama
Comedy
Foreign
Drama
Thriller
Family
Comedy
Drama
Drama
Romance
Drama
Comedy
Drama
Romance
Drama
History
Drama
Mystery
Comedy
Action
Thriller
Action
Drama
Comedy
Drama
Comedy
Romance
Drama
Romance
Crime
Drama
Romance
Comedy
Action
Comedy
Science Fiction
Science Fiction
Comedy
Drama
Horror
Romance
Thriller
Animation
Science Fiction
Drama
Horror
Horror
Drama
Family
Crime
Drama
Thriller
Action
Drama
Science Fiction
Thriller
Horror
Drama
Romance
Action
Comedy
Horror
Western
Drama
Foreign
Comedy
Action
Thriller
Crime
Drama
History
War
Western
Drama
Action
Comedy
Thriller
Drama
Drama
Action
Drama
Action
Action
Thriller
Action
Drama
Thriller
Adventur

Family
Drama
Comedy
Action
Adventure
Comedy
Crime
Drama
Thriller
Comedy
Family
Romance
Action
Comedy
Drama
Romance
Drama
Romance
Documentary
Action
Thriller
War
Western
Horror
Mystery
Thriller
Comedy
Drama
Romance
Drama
Drama
Foreign
Romance
Horror
Science Fiction
Thriller
Drama
Crime
Comedy
Action
Crime
Drama
Thriller
Documentary
Drama
Drama
Thriller
Crime
Romance
Drama
Comedy
Music
Drama
Mystery
Thriller
Adventure
Drama
Documentary
Drama
Documentary
Comedy
History
War
Thriller
Crime
Drama
Comedy
Family
Adventure
Family
Fantasy
Adventure
Drama
Action
Drama
History
Romance
Comedy
Drama
Romance
Fantasy
Drama
Action
Thriller
Romance
Crime
Drama
Thriller
Comedy
Documentary
Drama
Thriller
Romance
History
Drama
Drama
Drama
Drama
Romance
War
Drama
History
Drama
Action
Horror
Mystery
Thriller
Drama
Drama
Comedy
Foreign
Comedy
Drama
Fantasy
Romance
Action
Comedy
Crime
Drama
Thriller
Drama
Romance
Documentary
Action
Crime
Thriller
Crime
Drama
Thriller
Comedy
Romance
Comedy
Drama
Romance
Drama
A

Comedy
Comedy
Drama
Romance
Drama
Horror
Drama
Action
Thriller
Crime
Thriller
Action
Comedy
Crime
Thriller
Drama
Foreign
Drama
Crime
Drama
Action
Thriller
Horror
Thriller
Horror
Horror
Crime
Drama
Romance
Comedy
Drama
Comedy
Romance
Horror
Science Fiction
Family
Comedy
Documentary
Comedy
Drama
Crime
Drama
Fantasy
Mystery
Thriller
Comedy
Action
Thriller
Fantasy
Action
Adventure
Drama
Romance
Comedy
Family
Drama
Western
Romance
Documentary
Foreign
Comedy
Drama
Romance
Action
Drama
History
Documentary
Comedy
Drama
Romance
Adventure
Fantasy
Animation
Science Fiction
Drama
Thriller
Thriller
Action
Horror
Comedy
Drama
Romance
Drama
Horror
Thriller
Drama
Comedy
Drama
Comedy
Romance
Drama
Romance
Animation
Drama
Mystery
Science Fiction
Thriller
Drama
Comedy
Drama
Romance
Action
Drama
War
Drama
Romance
Drama
Thriller
Animation
Adventure
Family
Drama
Fantasy
Horror
Drama
Crime
Drama
Action
Thriller
Foreign
Drama
Romance
Comedy
Romance
Drama
Family
Comedy
Music
Romance
Documentary
Foreign
Action


Comedy
Family
Horror
Drama
War
Drama
Drama
Science Fiction
Action
Drama
Science Fiction
Drama
Mystery
Thriller
Action
Comedy
Romance
Drama
Comedy
Drama
Romance
Drama
Comedy
Music
Romance
Drama
Documentary
Horror
Drama
Thriller
Comedy
Drama
Drama
Thriller
Music
Documentary
Music
Drama
Drama
Crime
Comedy
Family
Comedy
Romance
Thriller
Horror
Mystery
Drama
History
Romance
Action
Fantasy
Horror
War
Drama
History
Adventure
Romance
Thriller
Science Fiction
Animation
Comedy
Crime
Romance
Thriller
Drama
Action
Drama
Thriller
Adventure
Family
Fantasy
Comedy
Adventure
Drama
Action
Science Fiction
Action
Comedy
Western
Documentary
Action
Drama
Foreign
Comedy
Drama
History
Romance
Comedy
Drama
Action
Western
War
Drama
Drama
Drama
Mystery
Thriller
Adventure
Drama
Comedy
Fantasy
Drama
Drama
Thriller
Comedy
Fantasy
Documentary
Crime
Drama
Action
Science Fiction
Comedy
Drama
Family
Romance
Comedy
Thriller
Drama
Romance
Animation
Family
Adventure
Drama
Fantasy
Horror
Science Fiction
Thriller
Drama
Roma

Horror
Mystery
Thriller
Action
Crime
Mystery
Thriller
Comedy
Drama
Drama
Romance
Drama
Romance
Adventure
Comedy
Drama
Comedy
Drama
History
History
Drama
Crime
Drama
Documentary
Foreign
Comedy
Documentary
Crime
Drama
Comedy
Comedy
Romance
Comedy
Music
Drama
Romance
Drama
Drama
Comedy
Romance
Action
Thriller
Romance
Drama
Drama
Romance
Drama
Romance
Romance
Horror
Thriller
Romance
Drama
Comedy
Comedy
Romance
Animation
Music
Drama
Drama
Romance
Comedy
Drama
Romance
Animation
Comedy
Music
Comedy
Science Fiction
Documentary
Comedy
Drama
Horror
Drama
Drama
Mystery
Thriller
Fantasy
Music
Romance
Adventure
Drama
Family
Fantasy
Drama
Music
Drama
Music
Drama
Romance
Drama
Fantasy
Documentary
History
Documentary
History
Music
Action
Drama
War
Comedy
War
Crime
Drama
Drama
Romance
Comedy
Drama
Drama
Action
Adventure
Drama
Action
History
Thriller
Documentary
Comedy
Family
Comedy
Drama
Romance
Drama
Crime
Thriller
Adventure
Romance
Science Fiction
Horror
Drama
Drama
Comedy
Action
Adventure
Drama
Fant

Drama
Horror
Thriller
Mystery
Romance
Foreign
Comedy
Drama
Drama
Horror
Horror
Horror
Crime
Thriller
Drama
Drama
Romance
Thriller
Action
Animation
Comedy
Fantasy
Mystery
Comedy
Drama
Music
Comedy
Comedy
Family
Fantasy
Action
Mystery
Thriller
Drama
Documentary
Comedy
Documentary
Crime
Drama
Thriller
Comedy
Drama
Romance
Drama
Music
Mystery
Comedy
Drama
Romance
Drama
Drama
Action
Science Fiction
Thriller
Documentary
Music
Music
Science Fiction
Drama
Romance
Fantasy
Comedy
Romance
Drama
War
Adventure
Drama
War
Drama
Romance
Comedy
Documentary
Drama
Music
Comedy
Drama
Romance
Drama
Horror
Thriller
Comedy
Documentary
Drama
Drama
Comedy
Comedy
Fantasy
Drama
Comedy
Foreign
Comedy
Drama
Thriller
Drama
Drama
Romance
Foreign
Music
Comedy
Drama
Comedy
Action
Crime
Horror
Mystery
Thriller
Crime
Action
Drama
Foreign
Thriller
Drama
Comedy
Foreign
Action
Comedy
Thriller
Drama
Crime
Drama
Mystery
Thriller
Drama
Romance
Action
Adventure
Family
Comedy
Animation
Family
Action
Crime
Drama
Comedy
Drama
Rom

Comedy
Comedy
Drama
Romance
Comedy
Drama
Comedy
Romance
Romance
Drama
Adventure
Drama
Documentary
Drama
Comedy
Drama
Documentary
Family
Fantasy
Thriller
Horror
Crime
Comedy
Action
Comedy
Action
Drama
Comedy
Music
Romance
Fantasy
Horror
Thriller
Science Fiction
Mystery
Drama
Romance
Drama
Romance
Comedy
Crime
Thriller
Drama
Drama
Romance
Documentary
Action
Fantasy
Thriller
Horror
Drama
Drama
Thriller
War
Thriller
Drama
Romance
Drama
Documentary
Drama
Drama
Action
Comedy
Thriller
Science Fiction
Comedy
Romance
Horror
Thriller
Drama
Crime
Action
Crime
Thriller
Animation
Family
Drama
Thriller
Drama
Thriller
Drama
War
History
Comedy
Thriller
Horror
Mystery
Thriller
Documentary
History
Drama
Romance
Foreign
Drama
Romance
Drama
Comedy
Documentary
Animation
Family
Drama
Drama
Drama
Comedy
Drama
History
War
Action
Adventure
Drama
Mystery
Thriller
Drama
Thriller
Documentary
Fantasy
Thriller
Mystery
Documentary
Horror
Thriller
Drama
Western
Comedy
Drama
Romance
Drama
Action
Comedy
Thriller
Crime


Romance
Science Fiction
Action
Thriller
Foreign
Drama
Music
Drama
Comedy
Horror
Horror
Action
Crime
Drama
Thriller
Drama
Romance
War
Drama
History
Action
Thriller
Drama
Drama
Thriller
Adventure
Action
Crime
Mystery
Documentary
Drama
Comedy
Foreign
Drama
Adventure
Comedy
Crime
Drama
Drama
Thriller
Drama
Western
Drama
Comedy
Drama
Drama
Drama
Documentary
Documentary
Comedy
Drama
Foreign
Foreign
Drama
Drama
Drama
Drama
Drama
Foreign
Drama
Comedy
Drama
Comedy
Drama
Comedy
Thriller
Mystery
Crime
Drama
Comedy
Romance
Action
Thriller
Adventure
Family
Fantasy
Action
Science Fiction
War
Drama
Comedy
Family
Comedy
Music
Romance
Thriller
Crime
Mystery
Drama
Horror
Action
Thriller
Science Fiction
Documentary
Comedy
Drama
Comedy
Fantasy
Family
Music
Animation
Drama
Horror
Thriller
History
Documentary
Comedy
Thriller
Action
Drama
Comedy
Drama
Romance
Drama
Romance
Horror
Thriller
Drama
Thriller
Crime
Drama
Drama
Romance
War
Horror
Drama
Crime
Drama
Comedy
Drama
Romance
Drama
Comedy
Drama
Drama
Roman

Vision View Entertainment
Telescene Film Group Productions
Drama
Comedy
Romance
Drama
War
Documentary
Drama
Romance
Adventure
Drama
Action
Mystery
Action
Drama
Horror
Science Fiction
Documentary
Action
Drama
Comedy
Documentary
Horror
Drama
Action
Horror
Thriller
Action
Horror
Comedy
Horror
Drama
Thriller
Drama
Horror
Science Fiction
Thriller
Drama
Crime
Drama
Thriller
Comedy
Drama
Music
Drama
Thriller
Drama
Drama
Comedy
Drama
Drama
Horror
Romance
Music
Drama
Animation
Science Fiction
Action
Action
Comedy
Thriller
Romance
Drama
Romance
Action
Drama
Thriller
Family
Horror
Thriller
Drama
Comedy
Documentary
Foreign
Drama
War
Comedy
Drama
Drama
Action
Crime
Drama
Thriller
Drama
Romance
Documentary
Crime
Drama
Documentary
Drama
Romance
Action
Drama
Foreign
Drama
Crime
Action
Thriller
Horror
Thriller
Drama
Horror
Thriller
Comedy
Documentary
Action
Comedy
Comedy
Horror
Drama
Romance
Horror
Drama
Thriller
Drama
Romance
Adventure
Horror
Action
Foreign
Action
Fantasy
Adventure
Fantasy
Horror
Docu

Thriller
Horror
Science Fiction
Horror
Science Fiction
Action
Adventure
Comedy
Fantasy
Action
Comedy
Crime
Drama
Fantasy
Science Fiction
Animation
Action
Adventure
Drama
History
Drama
Mystery
Action
Drama
Action
Comedy
Action
Thriller
Comedy
Foreign
Documentary
Science Fiction
Horror
Mystery
Science Fiction
Drama
History
Thriller
Horror
Thriller
Drama
Mystery
Drama
Science Fiction
Comedy
Action
Science Fiction
Horror
Action
Drama
Family
Drama
Drama
Thriller
Adventure
Fantasy
Drama
Family
Foreign
Comedy
Western
Documentary
Fantasy
Animation
Drama
War
Documentary
Documentary
Animation
Family
Western
Thriller
Action
Science Fiction
Adventure
Fantasy
Science Fiction
Action
Comedy
Drama
Romance
Documentary
Family
Adventure
Drama
Western
Documentary
Documentary
Drama
Foreign
Documentary
Romance
Drama
Drama
Comedy
Horror
Action
Comedy
Crime
Thriller
Action
Crime
Drama
Horror
Comedy
Action
Drama
Science Fiction
Crime
Mystery
Drama
Thriller
Drama
Comedy
Drama
Comedy
Drama
Fantasy
Romance
Scienc

Mystery
Romance
Drama
Drama
Mystery
Romance
Drama
Documentary
Drama
Action
Drama
Thriller
Crime
Action
Comedy
Action
Thriller
Drama
Romance
Drama
Animation
Family
Adventure
Comedy
Animation
Family
Comedy
Documentary
Horror
Documentary
Documentary
Comedy
Drama
Comedy
Drama
Horror
Family
Drama
Romance
Comedy
Foreign
Drama
Adventure
Comedy
Comedy
Animation
Fantasy
Drama
Drama
Fantasy
Science Fiction
Drama
Romance
Drama
Fantasy
Mystery
Romance
Comedy
Horror
Horror
Documentary
Drama
Documentary
Music
Adventure
Action
Science Fiction
Documentary
Action
Horror
Thriller
Fantasy
Drama
Romance
Drama
Adventure
Romance
Comedy
Action
Crime
Drama
Action
Thriller
Mystery
Foreign
Drama
Documentary
Comedy
Horror
Documentary
Music
Drama
Family
Action
Comedy
Western
Science Fiction
Romance
Adventure
Documentary
Documentary
Drama
Documentary
Comedy
Action
Science Fiction
Comedy
Adventure
Crime
Drama
Thriller
Comedy
Comedy
Comedy
Action
Adventure
Fantasy
Horror
Science Fiction
Fantasy
Animation
Documentary

Drama
War
Animation
Comedy
Family
Thriller
Horror
Horror
Drama
Fantasy
Horror
Thriller
Action
Crime
Comedy
Drama
Romance
Comedy
War
Drama
Thriller
Horror
War
Drama
Romance
Crime
Drama
Thriller
Action
Drama
Foreign
Action
Comedy
Animation
Romance
Science Fiction
Horror
Action
Drama
Fantasy
War
Crime
Drama
Music
Documentary
Comedy
Crime
Mystery
Thriller
Crime
Mystery
Comedy
Thriller
Comedy
Crime
Action
Comedy
Thriller
Comedy
Horror
Horror
Thriller
Horror
Mystery
Documentary
Science Fiction
Horror
Drama
Thriller
Science Fiction
Thriller
Action
Drama
Fantasy
Adventure
Family
Action
Drama
Mystery
Thriller
Crime
Drama
Mystery
Thriller
Science Fiction
Comedy
Science Fiction
Horror
Thriller
Horror
Drama
Drama
Romance
Drama
Drama
Thriller
Drama
Romance
Comedy
Drama
Thriller
Crime
Thriller
Drama
Science Fiction
Thriller
Animation
Romance
Adventure
Drama
Fantasy
Drama
Action
Animation
Comedy
Family
Drama
Music
Horror
Adventure
Animation
Family
Drama
Family
Drama
Horror
Science Fiction
Documentary

Comedy
Drama
Family
Horror
Comedy
Drama
Science Fiction
Thriller
Horror
Documentary
Comedy
Thriller
Mystery
Thriller
Horror
Adventure
Action
Family
Comedy
Action
Thriller
Science Fiction
Music
Romance
Romance
Thriller
Animation
Action
Comedy
Family
Thriller
Drama
Drama
Crime
Drama
Comedy
Romance
Drama
Romance
Action
Western
Western
Western
Drama
Comedy
Music
Drama
Romance
Crime
Drama
History
Drama
Action
Adventure
Drama
Western
Comedy
Music
Documentary
Documentary
Adventure
Thriller
Mystery
Comedy
Fantasy
Comedy
Romance
Comedy
Animation
Drama
Drama
Romance
Drama
Fantasy
Romance
Drama
Family
Drama
Drama
Romance
Comedy
Romance
Crime
Thriller
Comedy
Family
Science Fiction
Action
Drama
Science Fiction
Horror
Family
Comedy
Comedy
Documentary
Drama
History
Comedy
Drama
Science Fiction
Comedy
Drama
Romance
Comedy
Music
Action
Drama
Thriller
Action
Action
Thriller
Romance
Drama
Adventure
Fantasy
Drama
Comedy
Romance
Action
Science Fiction
Drama
Thriller
Action
Crime
Drama
Drama
Romance
Comedy


Adventure
Action
Adventure
Action
Comedy
Foreign
Comedy
Drama
Romance
Adventure
Fantasy
Romance
Science Fiction
Animation
Documentary
Science Fiction
Action
Thriller
Crime
Drama
Documentary
Horror
Science Fiction
Comedy
Drama
Science Fiction
Action
Animation
Adventure
Comedy
Comedy
Romance
Action
Drama
Western
Action
War
Romance
Thriller
Drama
Crime
Drama
Animation
Horror
Action
Crime
Thriller
Comedy
Romance
Comedy
Drama
Romance
Action
Western
Family
Animation
Adventure
Comedy
Drama
Animation
Drama
Romance
Thriller
Animation
Family
Fantasy
Drama
Action
Science Fiction
Thriller
Drama
Drama
Mystery
Comedy
Drama
Romance
Drama
Family
Western
Horror
Drama
Drama
History
Comedy
Action
Western
Romance
Drama
Crime
Drama
Mystery
Drama
Crime
Drama
Drama
Music
Romance
Animation
Family
Fantasy
Fantasy
Romance
Drama
Foreign
Adventure
Drama
Music
Drama
Drama
Comedy
Drama
Thriller
Crime
Adventure
Drama
Drama
Western
Action
Drama
Comedy
Comedy
Comedy
Comedy
Music
Romance
Drama
Comedy
Romance
Music
Dram

Drama
Action
Crime
Drama
Documentary
Drama
Documentary
Romance
Comedy
Drama
Comedy
Drama
Drama
War
Horror
War
Fantasy
Drama
Drama
Documentary
Drama
Family
Comedy
Comedy
Music
Documentary
Drama
Science Fiction
Thriller
Horror
Mystery
Thriller
Documentary
Crime
Action
Thriller
Comedy
Action
Science Fiction
Thriller
Drama
War
Action
Adventure
Horror
Science Fiction
Science Fiction
Comedy
Romance
Drama
Action
Adventure
Thriller
Drama
Fantasy
Animation
Drama
Romance
Action
Adventure
Science Fiction
Fantasy
Drama
Comedy
Animation
Adventure
Fantasy
Romance
Documentary
Horror
Mystery
Thriller
Action
Drama
Thriller
Crime
Drama
Crime
Thriller
Drama
Romance
Comedy
Crime
Drama
Romance
Action
Science Fiction
Horror
Science Fiction
Thriller
Science Fiction
Fantasy
Adventure
Family
Documentary
Horror
Thriller
Adventure
Family
Fantasy
Comedy
Drama
Mystery
Mystery
Drama
Drama
Thriller
History
Drama
Horror
Thriller
Documentary
Drama
Fantasy
Mystery
Thriller
Drama
Romance
War
Music
Documentary
Comedy
Dra

Thriller
Comedy
Animation
Drama
Drama
Comedy
Animation
Family
Drama
Documentary
Comedy
Action
Adventure
Comedy
Drama
Comedy
Romance
Foreign
Documentary
Mystery
Drama
Thriller
Drama
Action
Crime
Action
Drama
Thriller
Music
Drama
Documentary
Crime
Drama
Animation
Science Fiction
Thriller
Comedy
Comedy
Crime
Comedy
Western
Western
Horror
Thriller
Drama
History
Drama
Documentary
Documentary
Drama
History
Documentary
Music
Documentary
Drama
Comedy
Comedy
Music
Action
Crime
Drama
Thriller
Documentary
Drama
Comedy
Foreign
Documentary
Documentary
Documentary
Drama
Romance
Comedy
Comedy
History
Drama
Comedy
Drama
Comedy
Comedy
Documentary
Comedy
Adventure
Fantasy
Comedy
Comedy
Drama
Comedy
Documentary
Music
Comedy
Comedy
Drama
Fantasy
Horror
Comedy
Comedy
Documentary
Music
Drama
Fantasy
Comedy
Drama
Comedy
Comedy
Romance
Comedy
Documentary
Foreign
Comedy
Comedy
Comedy
Drama
Documentary
Romance
Family
Documentary
History
Action
Drama
Foreign
Documentary
Drama
Music
Romance
Drama
Romance
Action
T

Comedy
Horror
Western
Horror
Drama
Thriller
Drama
Comedy
Drama
Western
Drama
Adventure
Thriller
Drama
Romance
Comedy
Drama
Crime
Drama
Drama
Drama
Comedy
Drama
Romance
Drama
Mystery
Drama
Romance
Fantasy
Horror
Mystery
Comedy
Romance
Comedy
Animation
Drama
Crime
Fantasy
Action
Comedy
Science Fiction
Horror
Thriller
Drama
Action
Thriller
Science Fiction
Fantasy
Action
Horror
Horror
Comedy
Horror
Thriller
Horror
Action
Science Fiction
Foreign
Horror
Horror
Action
History
War
Drama
Animation
Drama
Fantasy
Action
Comedy
Crime
Foreign
Foreign
Horror
Mystery
Romance
Drama
Drama
Drama
Drama
Romance
Animation
Comedy
Comedy
Horror
Drama
Drama
Comedy
Romance
Foreign
Horror
Thriller
Horror
Drama
Action
War
Drama
Documentary
Horror
Thriller
Drama
Mystery
Thriller
Drama
Family
Comedy
Drama
Romance
Drama
Music
Comedy
Drama
Romance
Romance
Drama
Horror
Action
Crime
Foreign
Romance
Comedy
Drama
Drama
Comedy
Romance
Foreign
Action
Adventure
History
War
Comedy
Animation
Family
Family
Romance
Drama
Horro

Crime
Drama
Crime
Drama
Documentary
History
Documentary
Documentary
Documentary
Western
Western
Documentary
Drama
Comedy
Fantasy
Romance
Family
Comedy
Comedy
History
Drama
Comedy
Crime
Crime
Music
Crime
Drama
Action
Action
Adventure
Fantasy
Animation
Adventure
Comedy
Action
Family
Action
Comedy
Fantasy
Science Fiction
Thriller
Drama
Thriller
Science Fiction
Drama
History
Drama
History
Crime
Thriller
Horror
Documentary
Music
Horror
Thriller
Mystery
Action
Adventure
Foreign
Adventure
Drama
Thriller
Drama
Comedy
History
Music
Adventure
Family
Crime
Thriller
Drama
Drama
War
Drama
Comedy
Drama
Romance
Thriller
Comedy
Drama
Mystery
Thriller
Drama
Documentary
Romance
Drama
Family
Comedy
Science Fiction
Drama
Romance
Action
Drama
Thriller
Family
Fantasy
Drama
Comedy
Drama
Romance
Animation
Documentary
Crime
Drama
Mystery
Thriller
Drama
Action
Drama
Foreign
Romance
Thriller
Drama
Action
Comedy
Crime
Thriller
Drama
Mystery
Documentary
Drama
Comedy
Horror
Documentary
Thriller
Action
Crime
Action


Drama
Drama
Horror
Drama
Comedy
Drama
Fantasy
Action
Comedy
Science Fiction
Comedy
Drama
Romance
Crime
Comedy
Action
Adventure
Music
Romance
Comedy
Music
Romance
Drama
Mystery
Thriller
Comedy
Comedy
Romance
Music
War
Drama
Crime
Drama
Drama
History
War
War
Drama
Horror
Documentary
Thriller
Documentary
Documentary
Comedy
Documentary
Documentary
Documentary
Documentary
Romance
Thriller
Drama
Mystery
Drama
Fantasy
Drama
Romance
Foreign
Documentary
Comedy
Foreign
Drama
Drama
Romance
Drama
Drama
Action
Adventure
Adventure
Science Fiction
Drama
Mystery
Horror
Drama
Drama
Comedy
Crime
Drama
Comedy
Music
Romance
Comedy
Comedy
Drama
Romance
Family
Animation
Music
Animation
Music
Horror
Thriller
Documentary
Music
Drama
Mystery
Horror
Drama
Western
Action
Crime
Drama
Romance
Western
Action
Thriller
Drama
Horror
Action
Horror
Science Fiction
Documentary
Horror
Documentary
Action
Comedy
Drama
Comedy
Romance
Drama
Adventure
Science Fiction
Comedy
Family
Action
Thriller
Foreign
Western
Action
Thrille

Horror
Science Fiction
Thriller
Adventure
Documentary
Thriller
Adventure
Thriller
Drama
Drama
Thriller
Action
Thriller
Drama
Horror
Drama
Comedy
Romance
Romance
Drama
Action
Action
Thriller
Foreign
Thriller
Action
Romance
Drama
Foreign
Crime
Horror
Documentary
Foreign
Horror
Foreign
Drama
Action
Drama
Crime
Thriller
Action
Crime
Drama
Action
Drama
Family
Comedy
Documentary
Animation
Drama
Foreign
Romance
Action
Thriller
Drama
Romance
Drama
Romance
Foreign
Music
Mystery
Thriller
Horror
Drama
Action
Romance
Foreign
Action
Drama
Drama
Action
Foreign
Adventure
Action
Drama
Drama
Foreign
Drama
Foreign
Action
Drama
Romance
Drama
Action
Drama
Foreign
Romance
Thriller
Foreign
Documentary
Documentary
Documentary
Documentary
Documentary
Documentary
Romance
Drama
Comedy
Comedy
Thriller
Comedy
Animation
Adventure
Drama
Romance
Comedy
Drama
Music
Documentary
Documentary
Music
Music
Documentary
Crime
Documentary
Action
Fantasy
Comedy
Drama
Comedy
Adventure
Drama
Adventure
Comedy
Science Fiction
Acti

Comedy
Drama
Action
Comedy
Foreign
Drama
Drama
Comedy
Romance
Drama
Comedy
Drama
Action
Action
Documentary
Music
Comedy
Drama
Drama
Mystery
Thriller
Science Fiction
Thriller
Drama
Adventure
Action
Thriller
Family
Drama
Comedy
Comedy
Drama
Drama
Animation
Drama
Action
Thriller
Foreign
Fantasy
Animation
Science Fiction
Foreign
Western
Drama
Thriller
Thriller
Action
Animation
Science Fiction
Adventure
Crime
Drama
Adventure
Animation
Comedy
Fantasy
Drama
Crime
Action
Adventure
Animation
Action
Animation
Animation
Action
Science Fiction
Action
Animation
Fantasy
Science Fiction
Adventure
Animation
Comedy
Fantasy
Animation
Horror
Adventure
Animation
Comedy
Fantasy
Animation
Adventure
Comedy
Fantasy
Fantasy
Animation
Comedy
Adventure
Animation
Drama
Fantasy
Fantasy
Animation
Animation
Comedy
Fantasy
Animation
Science Fiction
Foreign
Action
Adventure
Animation
Science Fiction
Science Fiction
Animation
Action
War
Animation
Science Fiction
Animation
Action
Animation
Drama
Action
Comedy
Action
Ani

Thriller
Horror
Comedy
Action
Drama
Horror
Drama
Drama
Action
Crime
Thriller
Thriller
Romance
Drama
Comedy
Documentary
History
Comedy
Drama
Romance
Drama
Thriller
Romance
Drama
Comedy
Horror
Fantasy
Thriller
Mystery
Adventure
Action
Western
Drama
Action
Drama
War
Action
Comedy
Thriller
Action
Science Fiction
Drama
Romance
Comedy
Romance
Documentary
History
Documentary
Comedy
Drama
Action
Drama
Romance
Drama
Romance
Drama
Family
Romance
Western
Drama
Western
Drama
Action
Western
Family
Drama
Romance
Family
Drama
Western
Adventure
Animation
Family
Science Fiction
Action
Comedy
Action
Science Fiction
Documentary
Comedy
Comedy
Comedy
Drama
Romance
Animation
Adventure
Fantasy
Crime
Drama
Mystery
Thriller
Romance
Drama
Horror
Comedy
Drama
Comedy
Documentary
Action
Science Fiction
Thriller
Science Fiction
Action
Thriller
Animation
Science Fiction
Horror
Thriller
War
Drama
Romance
Drama
Animation
Horror
Thriller
Horror
Documentary
Documentary
Drama
Documentary
Drama
Horror
Mystery
Animation
Wa

Documentary
Science Fiction
Comedy
Drama
Adventure
Drama
Romance
Science Fiction
Documentary
Foreign
Drama
Drama
Drama
Comedy
Comedy
Horror
Mystery
Thriller
Thriller
Horror
Documentary
Horror
Action
Comedy
Documentary
Thriller
Comedy
Romance
Comedy
Romance
Foreign
Drama
Comedy
Thriller
Horror
Romance
Comedy
Documentary
Documentary
Drama
Documentary
Romance
Comedy
Drama
Drama
History
Drama
War
Comedy
Drama
Drama
Action
Foreign
Animation
Comedy
Adventure
Drama
Music
Comedy
Documentary
Documentary
Comedy
Crime
Drama
Thriller
Mystery
Comedy
Drama
Crime
Drama
Thriller
Romance
Drama
Thriller
Drama
Animation
Adventure
Animation
Comedy
Fantasy
Animation
Family
Comedy
Adventure
Drama
Animation
Comedy
Drama
Thriller
Comedy
Horror
Thriller
Horror
Horror
Drama
Action
Drama
History
Drama
Comedy
Drama
Crime
Drama
Crime
Thriller
Documentary
Crime
Drama
Mystery
Documentary
Animation
Comedy
Comedy
History
Documentary
Romance
Drama
Drama
Music
Documentary
Documentary
Horror
Drama
Comedy
Music
Crime
Dram

Adventure
Animation
Comedy
Family
Science Fiction
Drama
Romance
Comedy
Family
Drama
Adventure
Family
Thriller
Drama
Crime
Comedy
Adventure
Mystery
Drama
Science Fiction
Drama
Comedy
Action
Fantasy
Adventure
Drama
History
Drama
Action
Comedy
Drama
Thriller
Action
Comedy
Romance
Science Fiction
Thriller
Drama
Music
Animation
Thriller
Action
Adventure
Comedy
Drama
Music
Adventure
Action
Horror
Comedy
Family
Fantasy
Romance
Comedy
Documentary
Animation
Drama
Fantasy
Animation
Adventure
Comedy
Drama
Animation
Drama
Family
Drama
Romance
Drama
Comedy
Documentary
Documentary
Mystery
Thriller
Animation
Action
Crime
Thriller
Comedy
Drama
Drama
Action
Comedy
Documentary
Drama
Mystery
Comedy
Drama
Drama
Action
History
Science Fiction
Drama
Comedy
Action
Drama
Action
Adventure
Crime
Drama
Thriller
Animation
Drama
Romance
Documentary
Documentary
Drama
Horror
Science Fiction
Thriller
Romance
Crime
Drama
Mystery
Thriller
Horror
Drama
Drama
Horror
Thriller
Comedy
Comedy
Fantasy
Romance
Thriller
Horror


Comedy
Adventure
Western
Action
Drama
Comedy
Action
Thriller
Thriller
Documentary
Thriller
Horror
Documentary
Comedy
Horror
Animation
Horror
Science Fiction
Horror
Fantasy
Animation
Horror
Horror
Animation
Comedy
Science Fiction
Comedy
Romance
Fantasy
Drama
Crime
Drama
Drama
Science Fiction
Romance
Thriller
Horror
Western
Comedy
Drama
Family
Thriller
Crime
Comedy
Adventure
Animation
Horror
Horror
Mystery
Thriller
Family
Animation
Comedy
Drama
Horror
Documentary
Documentary
Fantasy
Animation
Science Fiction
Family
Animation
Adventure
Fantasy
Animation
Action
Family
Animation
Family
Comedy
Family
Comedy
Drama
Family
Animation
Family
Adventure
Fantasy
Animation
Science Fiction
Family
Adventure
Animation
Family
Fantasy
Animation
Family
Science Fiction
Animation
Comedy
Documentary
Romance
Comedy
Foreign
Drama
Foreign
Documentary
Comedy
Comedy
Comedy
Drama
Thriller
Action
Science Fiction
Thriller
Adventure
Drama
Comedy
Documentary
Fantasy
Mystery
Thriller
Comedy
Comedy
Drama
Comedy
Comedy
Co

In [27]:
df_info_pelis["generos"].apply(lambda x:len(x)).value_counts()

1    14814
2    14484
3     9493
4     3260
0     2449
5      789
6      151
7       23
8        3
Name: generos, dtype: int64

In [49]:
df["movieId"].value_counts()

356       341
296       324
318       311
593       304
260       291
         ... 
98604       1
103659      1
104419      1
115927      1
6425        1
Name: movieId, Length: 9066, dtype: int64

In [50]:
df["userId"].value_counts()

547    2391
564    1868
624    1735
15     1700
73     1610
       ... 
296      20
289      20
249      20
221      20
1        20
Name: userId, Length: 671, dtype: int64

In [10]:
df_entrenar=df.pivot_table(values="rating",index="userId",columns="movieId")


In [11]:
df_entrenar.head()

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     4.0     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      ...  161084  161155  161594  161830  161918  161944  \
userId                   ...                                                   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     4.0  ...     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     4.0  ...     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  162376  162542  162672  163949  
userId                                   
1           NaN     NaN     NaN     NaN  
2           NaN     NaN     NaN     NaN  
3           NaN     NaN     NaN     NaN  
4           NaN     NaN     NaN     NaN  
5           NaN     NaN     NaN     NaN  

[5 rows x 9066 columns]

In [12]:
from sklearn.decomposition import NMF
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity

In [13]:
df_entrenar2=df_entrenar.fillna(0)

In [14]:
df_entrenar2.head()

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
5           0.0     0.0     4.0     0.0     0.0     0.0     0.0     0.0   

movieId  9       10      ...  161084  161155  161594  161830  161918  161944  \
userId                   ...                                                   
1           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
2           0.0     4.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
3           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
4           0.0     4.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
5           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

movieId  162376  162542  162672  163949  
userId                                   
1           0.0     0.0     0.0     0.0  
2           0.0     0.0     0.0     0.0  
3           0.0     0.0     0.0     0.0  
4           0.0     0.0     0.0     0.0  
5           0.0     0.0     0.0     0.0  

[5 rows x 9066 columns]

In [15]:
distancias=cosine_similarity(df_entrenar2)

In [16]:
distancias[0]

array([1.        , 0.        , 0.        , 0.07448245, 0.01681799,
       0.        , 0.08388416, 0.        , 0.01284289, 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.05938727,
       0.        , 0.02107287, 0.        , 0.08563118, 0.01656352,
       0.05733056, 0.0436822 , 0.07024269, 0.        , 0.01491541,
       0.00262104, 0.03501505, 0.        , 0.        , 0.06130103,
       0.02379292, 0.03170691, 0.00821267, 0.08459736, 0.13058496,
       0.01652849, 0.        , 0.03282542, 0.0203909 , 0.        ,
       0.06833225, 0.        , 0.04105418, 0.        , 0.        ,
       0.        , 0.        , 0.02081219, 0.07333877, 0.        ,
       0.        , 0.        , 0.        , 0.03782836, 0.        ,
       0.03844173, 0.06430548, 0.        , 0.01381982, 0.        ,
       0.01399113, 0.        , 0.        , 0.        , 0.        ,
       0.04528371, 0.        , 0.00203034, 0.        , 0.03191959,
       0.        , 0.        , 0.08815485, 0.        , 0.04607

In [17]:
distancias=np.where(distancias==1,-1,distancias)

In [18]:
np.argpartition(distancias[0],-4)[-5:]

array([484, 309, 633, 324, 340], dtype=int64)

In [19]:
def devuelve_similares(indice,total_similares=5,distancias=distancias):
    return np.argpartition(distancias[indice],-total_similares)[-total_similares:]

Metamos, por ejemplo, al consumidor número 50:

In [20]:
similares_50=devuelve_similares(indice=50)

In [21]:
similares_50

array([655, 475, 625, 668, 188], dtype=int64)

Veamos ahora las películas que le gustan a nuestro 50:

In [22]:
df.head()

userId  movieId  rating
0       1       31     2.5
1       1     1029     3.0
2       1     1061     3.0
3       1     1129     2.0
4       1     1172     4.0

In [23]:
df_puntuaciones_50=df[df["userId"]==50][["movieId","rating"]]
df_puntuaciones_50

movieId  rating
7987       10     4.0
7988       21     3.0
7989       39     2.0
7990       47     3.0
7991       95     3.0
7992      110     4.0
7993      150     3.0
7994      160     3.0
7995      161     4.0
7996      165     4.0
7997      185     3.0
7998      208     3.0
7999      231     1.0
8000      253     3.0
8001      282     3.0
8002      292     4.0
8003      296     4.0
8004      315     3.0
8005      316     3.0
8006      337     3.0
8007      339     4.0
8008      344     2.0
8009      349     3.0
8010      356     4.0
8011      357     4.0
8012      367     4.0
8013      368     4.0
8014      377     3.0
8015      380     3.0
8016      420     3.0
8017      434     3.0
8018      440     4.0
8019      442     3.0
8020      454     3.0
8021      457     3.0
8022      480     4.0
8023      509     3.0
8024      527     4.0
8025      539     3.0
8026      553     3.0
8027      587     3.0
8028      589     5.0
8029      590     3.0
8030      597     3.0
8031      780     4.0
8032      786     3.0

In [24]:
df_info_pelis[df_info_pelis["id"].isin(df[df["userId"]==50]["movieId"].values)]

Empty DataFrame
Columns: [adult, belongs_to_collection, budget, genres, homepage, id, imdb_id, original_language, original_title, overview, popularity, poster_path, production_companies, production_countries, release_date, revenue, runtime, spoken_languages, status, tagline, title, video, vote_average, vote_count]
Index: []

[0 rows x 24 columns]

In [25]:
df_info_pelis["id"].dtype

dtype('O')

In [26]:
df["movieId"].dtype

dtype('int64')

In [27]:
df_info_pelis["id"].unique()

array(['862', '8844', '15602', ..., '67758', '227506', '461257'],
      dtype=object)

In [28]:
df["movieId"]=df["movieId"].astype("str")

In [29]:
df_pelis_50=df_info_pelis[df_info_pelis["id"].isin(df[df["userId"]==50]["movieId"].values)]
df_pelis_50

adult                              belongs_to_collection     budget  \
286    False  {'id': 300546, 'name': 'Once were Warriors Col...          0   
302    False  {'id': 131, 'name': 'Three Colors Collection',...          0   
385    False                                                NaN      45000   
1029   False                                                NaN   14500000   
1163   False                                                NaN    2200000   
1176   False  {'id': 119674, 'name': 'Psycho Collection', 'p...     806948   
1234   False                                                NaN    3500000   
1299   False  {'id': 8581, 'name': 'A Nightmare on Elm Stree...    1800000   
1639   False                                                NaN  200000000   
1808   False                                                NaN  140000000   
1852   False                                                NaN   25000000   
1902   False  {'id': 264, 'name': 'Back to the Future Collec...   40000000   
2556   False                                                NaN   42000000   
2875   False  {'id': 645, 'name': 'James Bond Collection', '...    7000000   
3531   False  {'id': 252313, 'name': 'The Endless Summer Col...          0   
3773   False                                                NaN   60000000   
4833   False  {'id': 304, 'name': "Ocean's Collection", 'pos...   85000000   
4919   False  {'id': 93295, 'name': '48 Hrs. Collection', 'p...   12000000   
5004   False                                                NaN          0   
5854   False                                                NaN   25000000   
6388   False  {'id': 528, 'name': 'The Terminator Collection...  200000000   
6833   False                                                NaN          0   
6994   False                                                NaN   70000000   
7146   False                                                NaN          0   
7218   False  {'id': 1952, 'name': 'USA: Land of Opportuniti...   10000000   
7600   False                                                NaN     600000   
9138   False  {'id': 344936, 'name': 'Unna Trilogie', 'poste...    2888030   
10395  False                                                NaN          0   
10538  False                                                NaN   50000000   
10781  False                                                NaN    3000000   
11912  False                                                NaN    2224000   
12284  False  {'id': 115762, 'name': 'AVP Collection', 'post...   40000000   
13443  False                                                NaN          0   
20013  False                                                NaN          0   
33216  False  {'id': 456, 'name': 'Sissi Collection', 'poste...          0   

                                                  genres  \
286                        [{'id': 18, 'name': 'Drama'}]   
302    [{'id': 18, 'name': 'Drama'}, {'id': 9648, 'na...   
385    [{'id': 28, 'name': 'Action'}, {'id': 80, 'nam...   
1029   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
1163   [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
1176   [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...   
1234   [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
1299                      [{'id': 27, 'name': 'Horror'}]   
1639   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
1808   [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
1852                       [{'id': 18, 'name': 'Drama'}]   
1902   [{'id': 12, 'name': 'Adventure'}, {'id': 35, '...   
2556   [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
2875   [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
3531                 [{'id': 99, 'name': 'Documentary'}]   
3773   [{'id': 18, 'name': 'Drama'}, {'id': 10402, 'n...   
4833   [{'id': 53, 'name': 'Thriller'}, {'id': 80, 'n...   
4919   [{'id': 53, 'name': 'Thriller'}, {'id': 28, 'n...   
5004   [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  

LLevamos las puntuaciones al df anterior:

In [30]:
df_pelis_50=df_pelis_50.merge(df_puntuaciones_50,left_on="id", right_on="movieId")[["title","id","rating"]]
df_pelis_50

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

Metemos todos los pasos anteriores en funciones:

In [32]:

##Función que obtiene las películas para un índice concreto:
def filtro_por_Id(indice,df_info_pelis=df_info_pelis,df=df):
    return df_info_pelis[df_info_pelis["id"].isin(df[df["userId"]==indice]["movieId"].values)][["title","id"]]

##Función para añadir al filtrado las puntuaciones. El valor de limitando indica
## si hay que limitar puntuaciones. Esta función internamnete llamará a la de arriba:

def anade_puntuaciones(indice,df=df,limitando = 0):
    df_salida=df[df["userId"]==50][["movieId","rating"]]
    df_salida=filtro_por_Id(indice=indice).merge(df_salida,left_on="id", right_on="movieId")[["title","id","rating"]]
    df_salida=df_salida[df_salida["rating"]>=limitando]
    df_salida["indice"]=indice
    return df_salida

##Replicamos aquí la función que obtiene los más cercanos:

def devuelve_similares1(indice,total_similares=5,distancias=distancias):
    return np.argpartition(distancias[indice],-total_similares)[-total_similares:]

###Aquí aplicamos ahora el proceso completo que irá llamando a las funciones de arriba:

def saca_recomendaciones(indice,df=df,total_similares =5,max_rating=4,filtro_solo_rec=False):
    ###Paso 1: Sacamos los argumentos asociados
    indices_aux=devuelve_similares1(indice,total_similares)
    ###Paso 2_ Sacamos las puntuaciens y películas del índicie
    pelis_salida=anade_puntuaciones(indice,limitando=0)
    ###Paso 3_ Añadimos las películas asociadas a los índices:
    lista_dfs=[anade_puntuaciones(x, limitando=max_rating) for x in indices_aux]
    for comentarista in lista_dfs:
        try:
            saca_indice=comentarista["indice"].values[0]
            pelis_salida=pelis_salida.merge(comentarista, on="id",suffixes=(None,saca_indice),how="outer")
        except:
            pass
    ##El siguiente filtro se quedará solo con las recomendaciones:
    if filtro_solo_rec:
        pelis_salida=pelis_salida[pelis_salida["title"].isnull()]
    #pelis_salida=pd.concat([pelis_salida]+lista_dfs,axis=1)
    return pelis_salida


In [33]:
saca_recomendaciones(475,filtro_solo_rec=False)

title   id  rating  indice  \
0                    Three Colors: Red  110     4.0   475.0   
1                      Monsoon Wedding  480     4.0   475.0   
2   Terminator 3: Rise of the Machines  296     4.0   475.0   
3                          Grill Point  316     3.0   475.0   
4                                  NaN  527     NaN     NaN   
5                                  NaN  339     NaN     NaN   
6                                  NaN  165     NaN     NaN   
7                                  NaN  780     NaN     NaN   
8                                  NaN  292     NaN     NaN   
9                                  NaN  161     NaN     NaN   
10                                 NaN  440     NaN     NaN   

                              title188  rating188  indice188  \
0                    Three Colors: Red        4.0      188.0   
1                      Monsoon Wedding        4.0      188.0   
2   Terminator 3: Rise of the Machines        4.0      188.0   
3                                  NaN        NaN        NaN   
4                   Once Were Warriors        4.0      188.0   
5                       Night on Earth        4.0      188.0   
6           Back to the Future Part II        4.0      188.0   
7           The Passion of Joan of Arc        4.0      188.0   
8         Dave Chappelle's Block Party        4.0      188.0   
9                                  NaN        NaN        NaN   
10                                 NaN        NaN        NaN   

                      title189  rating189  indice189            title625  \
0                          NaN        NaN        NaN                 NaN   
1                          NaN        NaN        NaN                 NaN   
2                          NaN        NaN        NaN                 NaN   
3                          NaN        NaN        NaN                 NaN   
4                          NaN        NaN        NaN  Once Were Warriors   
5                          NaN        NaN        NaN                 NaN   
6                          NaN        NaN        NaN                 NaN   
7   The Passion of Joan of Arc        4.0      189.0                 NaN   
8                          NaN        NaN        NaN                 NaN   
9                          NaN        NaN        NaN                 NaN   
10                         NaN        NaN        NaN                 NaN   

    rating625  indice625                            title385  rating385  \
0         NaN        NaN                   Three Colors: Red        4.0   
1         NaN        NaN                     Monsoon Wedding        4.0   
2         NaN        NaN  Terminator 3: Rise of the Machines        4.0   
3         NaN        NaN                                 NaN        NaN   
4         4.0      625.0                  Once Were Warriors        4.0   
5         NaN        NaN                      Night on Earth        4.0   
6         NaN        NaN          Back to the Future Part II        4.0   
7         NaN        NaN                                 NaN        NaN   
8         NaN        NaN        Dave Chappelle's Block Party        4.0   
9         NaN        NaN                      Ocean's Eleven        4.0   
10        NaN        NaN         Aliens vs Predator: Requiem        4.0   

    indice385  
0       385.0  
1       385.0  
2       385.0  
3         NaN  
4       385.0  
5       385.0  
6       385.0  
7         NaN  
8       385.0  
9       385.0  
10      385.0

In [123]:
devuelve_similares1(475)

array([188, 189, 625, 385, 540], dtype=int64)

Añadimos las películas de la recomendación:

In [126]:
anade_puntuaciones(188)

title   id  rating  indice
0                    Once Were Warriors  527     4.0     188
1                     Three Colors: Red  110     4.0     188
2                        Romeo + Juliet  454     3.0     188
3                                Psycho  539     3.0     188
4                        Night on Earth  339     4.0     188
5             A Nightmare on Elm Street  377     3.0     188
6                               Titanic  597     3.0     188
7                              Rain Man  380     3.0     188
8            Back to the Future Part II  165     4.0     188
9                          Notting Hill  509     3.0     188
10                     Live and Let Die  253     3.0     188
11                              48 Hrs.  150     3.0     188
12                      Monsoon Wedding  480     4.0     188
13                            The Hours  590     3.0     188
14   Terminator 3: Rise of the Machines  296     4.0     188
15           The Passion of Joan of Arc  780     4.0     188
16                             Big Fish  587     3.0     188
17                          Grill Point  316     3.0     188
18                     Bang, Boom, Bang  344     2.0     188
19                  Cockles and Muscles  349     3.0     188
20                              Syriana  231     1.0     188
21         Dave Chappelle's Block Party  292     4.0     188
22  The Arrival of a Train at La Ciotat  160     3.0     188
23                                Sissi  457     3.0     188

In [48]:
 modelo = NMF(n_components=2, init='random', random_state=0)

In [12]:
df_entrenar2=np.where(df_entrenar2==-1,1,df_entrenar2)

In [ ]:
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity

In [13]:
modelo.fit(df_entrenar2)

NMF(init='random', n_components=2, random_state=0)